metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -1119,6 +1053,62 @@ void validateManualPartitionAssignment(List<Integer> assignment,
         }
     }
 
+    void generateLeaderAndIsrUpdates(String context,
+                                     int brokerToRemoveFromIsr,
+                                     List<ApiMessageAndVersion> records,
+                                     Iterator<TopicIdPartition> iterator) {
+        int oldSize = records.size();
+        while (iterator.hasNext()) {
+            TopicIdPartition topicIdPart = iterator.next();
+            TopicControlInfo topic = topics.get(topicIdPart.topicId());
+            if (topic == null) {
+                throw new RuntimeException("Topic ID " + topicIdPart.topicId() + " existed in " +
+                    "isrMembers, but not in the topics map.");
+            }
+            PartitionControlInfo partition = topic.parts.get(topicIdPart.partitionId());
+            if (partition == null) {
+                throw new RuntimeException("Partition " + topicIdPart +
+                    " existed in isrMembers, but not in the partitions map.");
+            }
+            int[] newIsr = Replicas.copyWithout(partition.isr, brokerToRemoveFromIsr);
+            int newLeader = Replicas.contains(newIsr, partition.leader) ? partition.leader :
+                bestLeader(partition.replicas, newIsr, false);
+            boolean unclean = newLeader != NO_LEADER && !Replicas.contains(newIsr, newLeader);
+            if (unclean) {
+                // After an unclean leader election, the ISR is reset to just the new leader.
+                newIsr = new int[] {newLeader};
+            } else if (newIsr.length == 0) {
+                // We never want to shrink the ISR to size 0.
```

Comment:
The `newLeader` is `NO_LEADER` (due to empty `newIsr`) so we could set `NO_LEADER` to `record`. Does it make sense?

Reply:
I agree that if `newIsr.length` is 0 on line 1080, then `newLeader` must be `NO_LEADER`.  However, it's not necessary to add any special logic for this since `bestLeader` already handles this case.